pkg/controller/utils/clusterdeploymentbuilder.go

```diff
@@ -0,0 +1,111 @@
+package utils
+
+import (
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1alpha1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+type clusterDeploymentChecksumFunc func(clusterDeployment *hivev1.ClusterDeployment) (string, error)
+
+// ClusterDeploymentBuilder is a utility to build ClusterDeployments given specific requirements.
+// It's especialy useful in cases where simply defining a ClusterDeployment struct isn't sufficient
+// (e.g. when declaring a variable that needs to checksum attribute set.)
+type ClusterDeploymentBuilder struct {
+	ClusterDeployment *hivev1.ClusterDeployment
+	ChecksumFunc      clusterDeploymentChecksumFunc
+	Errors            []error
+}
+
+// NewClusterDeploymentBuilder is a constructor that creates a new ClusterDeploymentBuilder object given a specific checksum function.
+func NewClusterDeploymentBuilder(checksumFunc clusterDeploymentChecksumFunc) *ClusterDeploymentBuilder {
+	return NewClusterDeploymentBuilderWithClusterDeployment(checksumFunc, &hivev1.ClusterDeployment{})
+}
+
+// NewClusterDeploymentBuilderWithClusterDeployment is a constructor that creates a new ClusterDeploymentBuilder object
+// given a specific checksum function and an existing cluster deployment.
+func NewClusterDeploymentBuilderWithClusterDeployment(checksumFunc clusterDeploymentChecksumFunc, clusterDeployment *hivev1.ClusterDeployment) *ClusterDeploymentBuilder {
+	return &ClusterDeploymentBuilder{
+		ClusterDeployment: clusterDeployment,
+		ChecksumFunc:      checksumFunc,
+	}
+}
+
+// WithName sets the name of a cluster deployment
+func (n *ClusterDeploymentBuilder) WithName(name string) *ClusterDeploymentBuilder {
+	n.ClusterDeployment.Name = name
+	return n
+}
+
+// WithNamePostfix appends the given postfix to the name of the cluster deployment.
+func (n *ClusterDeploymentBuilder) WithNamePostfix(postfix string) *ClusterDeploymentBuilder {
+	n.ClusterDeployment.Name = n.ClusterDeployment.Name + "-" + postfix
+	return n
+}
+
+// WithNamespace sets the namespace of the cluster deployment
+func (n *ClusterDeploymentBuilder) WithNamespace(namespace string) *ClusterDeploymentBuilder {
+	n.ClusterDeployment.Namespace = namespace
+	return n
+}
+
+// WithBackupChecksum sets the backup checksum for the clusterdeployment using the preveiously set checksum function.
+func (n *ClusterDeploymentBuilder) WithBackupChecksum() *ClusterDeploymentBuilder {
```

Comment:
is there ever a case where you would normally do additional modifications to the object after calling WithBackupChecksum() (besides test cases)?

Reply:
Changes can definitely be made, but the checksum will then be incorrect.

This means that the object will be considered "changed since last backup".

Did that answer your question?  Are you asking for a change here?

Reply:
One thing you could do here to ensure that `WithBackupChecksum` is the last thing called in the chain is to have it return a different object.

```
func (n *ClusterDeploymentBuilder) WithBackupChecksum() *ClusterDeploymentChecksummedBuilder {
   ...
}

type ClusterDeploymentChecksummedBuilder struct {
  builder *ClusterDeployemntBuilder
}

func (n *ClusterDeploymentChecksummedBuilder) GenerateClusterDeployment (...) {
  return n.builder.GenerateClusterDeployment()
}
```

Reply:
no change requested. just making sure i understand what's going on.